storagenode/vouchers/service.go

```diff
@@ -7,16 +7,165 @@ import (
+
+	vdb   DB
+	trust *trust.Pool
+
+	expirationBuffer time.Duration
+
+	Loop sync2.Cycle
+}
+
+// NewService creates a new voucher service
+func NewService(log *zap.Logger, kad *kademlia.Kademlia, transport transport.Client, vdb DB, trust *trust.Pool, interval, expirationBuffer time.Duration) *Service {
+	return &Service{
+		log:              log,
+		kademlia:         kad,
+		transport:        transport,
+		vdb:              vdb,
+		trust:            trust,
+		expirationBuffer: expirationBuffer,
+		Loop:             *sync2.NewCycle(interval),
+	}
+}
+
+// Run sends requests to satellites for vouchers
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return service.Loop.Run(ctx, service.RunOnce)
+}
+
+// RunOnce runs one iteration of the voucher request service
+func (service *Service) RunOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("Checking vouchers")
+
+	satellites := service.trust.GetSatellites((ctx))
+
+	if len(satellites) > 0 {
+		var group errgroup.Group
+		ctx, cancel := context.WithTimeout(ctx, time.Hour)
+		defer cancel()
+		for _, sat := range satellites {
+			sat := sat
+			need, err := service.vdb.NeedVoucher(ctx, sat, service.expirationBuffer)
+			if err != nil {
+				service.log.Error("getting voucher status", zap.Error(err))
+				return nil
+			}
+			if need {
+				group.Go(func() error {
+					service.Request(ctx, sat)
+					return nil
+				})
+			}
+		}
+		_ = group.Wait() // doesn't return errors
+	} else {
+		service.log.Debug("No voucher requests to send")
+	}
+	return nil
+}
+
+// Request makes a voucher request to a satellite
+func (service *Service) Request(ctx context.Context, satelliteID storj.NodeID) {
+	service.log.Info("Requesting voucher", zap.String("satellite", satelliteID.String()))
+	err := service.request(ctx, satelliteID)
+	if err != nil {
+		service.log.Error("Error requesting voucher", zap.String("satellite", satelliteID.String()), zap.Error(err))
+	}
+}
+
+func (service *Service) request(ctx context.Context, satelliteID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	satellite, err := service.kademlia.FindNode(ctx, satelliteID)
+	if err != nil {
+		return VoucherError.New("unable to find satellite on the network: %v", err)
+	}
+
+	conn, err := service.transport.DialNode(ctx, &satellite)
+	if err != nil {
+		return VoucherError.New("unable to connect to the satellite: %v", err)
+	}
+	defer func() {
+		if cerr := conn.Close(); cerr != nil {
+			err = errs.Combine(err, VoucherError.New("failed to close connection: %v", err))
+		}
+	}()
+
+	resp, err := pb.NewVouchersClient(conn).Request(ctx, &pb.VoucherRequest{})
+	if err != nil {
+		return VoucherError.New("failed to start request: %v", err)
+	}
+
+	// TODO: separate status for disqualified nodes?
+	switch resp.GetStatus() {
+	case pb.VoucherResponse_REJECTED:
+		service.log.Info("Voucher request denied. Node does not meet voucher requirements")
```

Comment:
is there anyway we can provide more info about why it was rejected? Even if we can pass the specific values, but just generally what is being checked